### ISSUE TITLE
Get access-control-allow-origin for API

### ISSUE THREAD

**Issue by timothyfcook** [timothyfcook/pgh_col#5]
So we can pull live data from Chicago site for prototyping. Or figure out how to get Pgh API data.

**Comment by timothyfcook** [comment]
@teverette Thanks for the Pgh API info, didn't know what the engine URL was. Good to have.

What about setting `Access-Control-Allow-Origin: *` so we can use the data for this widget we are building?

**Comment by timothyfcook** [comment]
Just marking this here for the future. The Pittsburgh col-engine API address is: http://pitt.col-engine-staging.com/api/v1/orgs.json

... now we just need to have access so we can grab the data into our script instead of just using the static file.

**Comment by teverette** [comment]
Hi Tim, can you explain a bit more how you are using the api via Javascript. The concern is that if you are doing on the client side, are you making the secret public?

**Comment by MatthewVita** [comment]
This is unrelated to the secret (which wouldn't be persisted in the client
code anyways). This has to do with enabling CORS... run this in your
browser (check the console for the error):

```
var x =new XMLHttpRequest();
x.open("GET","http://pitt.col-engine-staging.com/api/v1/orgs.json",true);
x.send();
```

**Comment by timothyfcook** [comment]
Thanks @MatthewVita. Exactly: Aside, from grabbing static .json files, we cannot access the API currently due to the same-origin policy.

**Comment by teverette** [comment]
Hi Guys. The API calls are not really meant to be called from the client side in that manner. Very few of them will not require authentication. The secret might not be persisted but I'm not sure how you would call an api that requires authentication without encoding it on the server side. If you do that, then you might as well make the call from the server. I may not understand your need or issue but we can solve that part.

 However for your purposes of using orgs.json you can use jsonp to get passed the cross domain issue.
```html
<html>

<script>

function jsonp(url, callback) {
    var callbackName = 'jsonp_callback_' + Math.round(100000 * Math.random());
    window[callbackName] = function(data) {
        delete window[callbackName];
        document.getElementById('hidden').removeChild(script);
        callback(data);
    };

    var script = document.createElement('script');
    script.src = url + (url.indexOf('?') >= 0 ? '&' : '?') + 'callback=' + callbackName;
    document.getElementById('hidden').appendChild(script);
}
var load_orgs = function() {
  jsonp('http://pitt.col-engine-staging.com/api/v1/orgs.json', function(data) {
   alert("number of orgs: " + data.result.length);
});
};

</script>

<body onload="load_orgs();">

test cors

<div id="hidden" style="display:none;"></div>
</body>

</html>
```

**Comment by MatthewVita** [comment]
JSONP is a workaround in this case, but not a desirable one. JSONP allows for GETs, but not POSTs/DELETEs/PUTs, etc. CORS is, understandably, the solution in this setting.

A proper API should hold no opinions on client consumption and rendering. One should be able to write up any kind of client (Angular, Ember, Android, iOS, etc) to consume the API, presenting an auth token with requests when needed.

Good news is that I'm actually more of a back-end person and can help out in this space.

**Comment by timothyfcook** [comment]
Ok, is there anyway to not do this undesirable workaround? It seems pretty crucial for our work to be able to easily consume the API on the client side.

@teverette , is the Auth token you gave me just for the Partner API or for the City API as well? It's unclear to me how those are differentiated.

**Comment by timothyfcook** [comment]
@teverette what does it look like to work towards a solution here for client side API calls? Sounds like the JSONP workaround only works for GETS and is not really a solution. Thoughts?

**Comment by teverette** [comment]
@timothyfcook The auth token I provided is for the City API. As the city development team you don't necessarily need a partner api key, although Sprout Fund could be an org and could use or share their tokens with a third party for badging as an example.

**Comment by MatthewVita** [comment]
@teverette why are you guys handing out auth tokens... shouldn't these be unique granted per user auth?

**Comment by teverette** [comment]
@MatthewVita @timothyfcook I agree that APIs shouldn't necessarily dictate implementation. However, the platform is the access point to each city's data, and each city requires this data to be kept as secure as possible. 

That being said, I'd love to hear your plans for how you might implement that calls. Server side calls are already available to you all as an unlimited access point. If you plan on going through the client, I do think we can enable CORS for PGH with the caveat that you all don't expose your city's keys. (Clearly I don't think you would intend to do so but must be stated nonetheless.)

Maybe I can steal about 10 minutes of your time before the call Pitt has scheduled today with Mighel? Or some other time this afternoon?

**Comment by MatthewVita** [comment]
I'm a bit confused by "the platform is the access point to each city's data"... are you saying that each city API talks to a central API as opposed to separate forks?

**Comment by timothyfcook** [comment]
Thanks for the responses @teverette , I can chat for a bit ~3:00 EST before the other call. @MatthewVita @whit537 @justinxreese Any specific things I should talk to Tre about regarding the API, enabling CORS, etc.? Or, we could do a call on Friday when the team is all here working. Maybe both.

**Comment by teverette** [comment]
@MatthewVita I sent Tim the city's API token information that initially provided the only access into the system at the time. There are indeed per user tokens at the site (city) level that would be generated as needed.

Re: "the platform is the access point to each city's data" » I was just trying to convey that this is the access point to your city's data. 

@timothyfcook I can do ~3:00. Let me know what you all decide. Thanks!

**Comment by MatthewVita** [comment]
@teverette ohhhh, that clears this up. We definitely want to use the user auth generated tokens moving forward.

**Comment by timothyfcook** [comment]
@teverette able to talk now, Setup a hangout here to record it for other Devs to listen to later: https://plus.google.com/events/cqk7gfna5suae8ubqp62l0vgqjs

**Comment by timothyfcook** [comment]
@teverette sent you a note. Hoping to talk tomorrow sometime before or after lunch.
